Lesson_3/task_4.js

```diff
@@ -11,72 +11,93 @@
 
 
 const kupiBilet = function () {
-    let events = [];
+    let events = {};
     let cashier = 0;
-    let sold_tickets = [];
+    let sold_tickets = {};
 
     const generateId = function () {
         let generatedID = function () {
-           return  Array.from(Array(6), () => Math.floor(Math.random() * 36).toString(36)).join('');
-        }
-        if (generatedID() in sold_tickets.map(t => t.id)) {
-            return  generateId();
-        } else {
-            return generatedID();
+            return Array.from(Array(6), () => Math.floor(Math.random() * 36).toString(36)).join('');
         }
+
+        // recursive case
+        // let id = generatedID();
+        //
+        // return Boolean(sold_tickets[id]) ? generateId() : id;
+
+        // do/while case
+        let id;
+        do {
+            id = generatedID()
+        } while (Boolean(sold_tickets[id]))
+        return id;
     }
 
     const buyTicket = (event_name) => {
-       let event = events.find(e => e['event_name'] === event_name)
-       let ticket = function() {
-           if (event) {
-               cashier += event.ticket_price
-              let ticketObj = { id: generateId(), event_name: event.event_name, cost: event.ticket_price}
-               sold_tickets.push(ticketObj)
-               return ticketObj
-           } else {
-               return console.log('Event not found')
-           }
-       }
+        let event = events[event_name]
+        let ticket = function () {
+            if (event) {
+                cashier += event.ticket_price
+                let ticketObj = {id: generateId(), event_name: event.name, cost: event.ticket_price}
+
+                sold_tickets[ticketObj.id] = ticketObj
+                return ticketObj
+            } else {
+                return console.log('Event not found')
+            }
+        }
 
         return ticket();
```

Comment:
Предлагаю освободить немного памяти и улучшить читабельность кода, путем устранения вложенной функции:

```suggestion
            if (event) {
                cashier += event.ticket_price
                let ticketObj = {id: generateId(), event_name: event.name, cost: event.ticket_price}

                sold_tickets[ticketObj.id] = ticketObj
                return ticketObj
            } else {
                return console.log('Event not found')
            }
```